File: Algorithms/Searching/BinarySearch/binsearch.c
/* Binary Seach Algorithm 
 * e.g. input can be strings and ints */

#include <stdio.h>

int binary_search(int element, int startIndex, int endIndex, int arr[])
{
  int middleIndex;
  int result=-1;
  
  if (startIndex <= endIndex) {
    middleIndex = (startIndex + endIndex)/2;
    if (arr[middleIndex] == element)
      return middleIndex;
    if (element > arr[middleIndex])
      result = binary_search(element, middleIndex, endIndex, arr);
    else 
      result = binary_search(element, startIndex, middleIndex, arr);
  }
  return result;
}

int main(int argc, char *argv)
{
  int test_array[] = {1,2,3,4,5,6,7,8,9,10};
  int startIndex = 0;
  int endIndex = sizeof(test_array)/sizeof(int) - 1;
  int element = 4;
  int result = binary_search(element, startIndex, endIndex, test_array);
  
  if (result > -1)
    printf("Found element %d, found at position %d\n", element, result);
  else
    printf("No element found in array\n");

  return 0;
}
    

File: src/Searching/BinarySearch/binsearch.py
# Binary Seach in Python
# Runtime O(logn)
import unittest
from random import randint

class TestBinarySearch(unittest.TestCase):

    def setUp(self):
        self.test_array = sorted([randint(0, 100) for x in range(20)])
        self.element = self.test_array[randint(0,19)]

    def test_success(self):
        result = binary_search(self.element, self.test_array)
        self.assertTrue(self.test_array[result], self.element)
    
    def test_failure(self):
        notelement = 0
        while (notelement in self.test_array):
            notelement += 1
        result = binary_search(notelement, self.test_array)
        self.assertTrue(result, -1)


def binary_search(element, array):
    endIndex = len(array)-1
    startIndex = 0

    while startIndex < endIndex:
        middleIndex = (startIndex + endIndex)//2
        if array[middleIndex] == element:
            return middleIndex
        if element < array[middleIndex]:
            endIndex = middleIndex-1
        else:
            startIndex = middleIndex+1

    return -1


if __name__ == "__main__":
    unittest.main()
